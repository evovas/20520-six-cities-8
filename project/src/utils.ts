import {Offer} from './types/data';

export const calculateRatingStars = (rating: number): string => {
  const STARS_COUNT = 5;
  return `${Math.round(rating) * 100 / STARS_COUNT}%`;
};

export const spliceOffersByCity = (offers: Offer[]): {[key: string]: Offer[]} => (
  offers.reduce<{[city: string]: Offer[]}>((acc, offer)=> {
    const city = offer.city.name;
    if (!acc[city]) {
      acc[city] = [];
    }
    acc[city].push(offer);
    return acc;
  }, {})
);

export const getRandomPositiveInteger = (a: number, b:number): number => {
  const lower = Math.ceil(Math.min(Math.abs(a), Math.abs(b)));
  const upper = Math.floor(Math.max(Math.abs(a), Math.abs(b)));
  const result = Math.random() * (upper - lower + 1) + lower;

  return Math.floor(result);
};
