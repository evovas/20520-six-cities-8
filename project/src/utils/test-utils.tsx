import {Action, AnyAction, createStore, Store} from '@reduxjs/toolkit';
import {ReactNode} from 'react';
import {render, RenderResult} from '@testing-library/react';
import {rootReducer} from '../store/root-reducer';
import {Provider} from 'react-redux';
import {Router} from 'react-router-dom';
import {createMemoryHistory, MemoryHistory} from 'history';

interface RenderWithRedux<
  S = any,
  A extends Action = AnyAction,
  I extends S = any
  > {
  (
    ui: ReactNode,
    reduxOption?: {
      preloadedState?: I
      store?: Store<S, A>
    },
    memoryHistory?: {
      history?: MemoryHistory,
    }
  ) : RenderResult
}

const renderWithRedux: RenderWithRedux = (
  ui: ReactNode,
  {
    preloadedState,
    store = createStore(rootReducer, preloadedState),
  } = {},
  {
    history,
  } = {},
) => ({
  ...render(
    <Provider store={store}>
      <Router history={history ?? createMemoryHistory()}>
        {ui}
      </Router>
    </Provider>,
  ),
});

export * from '@testing-library/react';
export { renderWithRedux };
