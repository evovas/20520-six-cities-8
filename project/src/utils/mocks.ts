import {commerce, image, datatype, helpers, name, address, date, internet} from 'faker';
import {City, CurrentUser, Location, Offer, Review, User} from '../types/data';
import {getRandomPositiveInteger} from '../utils';
import dayjs from 'dayjs';

export const makeFakeUser = (): User => ({
  id: getRandomPositiveInteger(1, 100000),
  name: name.firstName(),
  avatarUrl: image.imageUrl(),
  isPro: datatype.boolean(),
});

export const makeFakeCurrentUser = (): CurrentUser => ({
  id: getRandomPositiveInteger(1, 100000),
  avatarUrl: image.imageUrl(),
  email: internet.email(),
  isPro: datatype.boolean(),
  name: name.firstName(),
  token: datatype.string(25),
});

export const makeFakeLocation = (): Location => ({
  latitude: +address.latitude(),
  longitude: +address.longitude(),
  zoom: getRandomPositiveInteger(8, 12),
});

export const makeFakeCity = (): City => ({
  location: makeFakeLocation(),
  name: address.city(),
});

export const makeFakeOffer = (): Offer => ({
  id: getRandomPositiveInteger(1, 10000000),
  bedrooms: getRandomPositiveInteger(1, 6),
  city: makeFakeCity(),
  description: commerce.productDescription(),
  goods: [commerce.productMaterial(), commerce.productMaterial(), commerce.productMaterial()],
  host: makeFakeUser(),
  images: [
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
  ],
  isFavorite: datatype.boolean(),
  isPremium: datatype.boolean(),
  location: makeFakeLocation(),
  maxAdults: getRandomPositiveInteger(1, 6),
  previewImage: image.imageUrl(),
  price: getRandomPositiveInteger(100, 1000),
  rating: getRandomPositiveInteger(10, 50) / 10,
  title: commerce.productName(),
  type: helpers.randomize(['apartment', 'room', 'house', 'hotel']),
});

export const makeFakeReview = (): Review => ({
  id: getRandomPositiveInteger(1, 10000000),
  comment: datatype.string(getRandomPositiveInteger(50, 300)),
  date: dayjs(date.past()).toISOString(),
  rating: getRandomPositiveInteger(1, 5),
  user: makeFakeUser(),
});
