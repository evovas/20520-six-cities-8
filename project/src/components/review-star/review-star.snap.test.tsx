import ReviewStar from './review-star';
import {renderWithRedux} from '../../utils/test-utils';

describe('Component: ReviewStar (snapshot)', () => {
  test('should render correctly', () => {
    const {container} = renderWithRedux(
      <ReviewStar
        value={'3'}
        currentValue={'3'}
        id={'3-stars'}
        onReviewChange={jest.fn()}
        disabled={false}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
