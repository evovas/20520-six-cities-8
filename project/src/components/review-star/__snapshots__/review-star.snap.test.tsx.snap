// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ReviewStar (snapshot) should render correctly 1`] = `
<div>
  <input
    checked=""
    class="form__rating-input visually-hidden"
    data-testid="review-star-3-stars"
    id="3-stars"
    name="rating"
    type="radio"
    value="3"
  />
  <label
    class="reviews__rating-label form__rating-label"
    for="3-stars"
    title="perfect"
  >
    <svg
      class="form__star-image"
      height="33"
      width="37"
    >
      <use
        xlink:href="#icon-star"
      />
    </svg>
  </label>
</div>
`;
